bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1075,15 +1075,24 @@ private boolean resolveConflict(LedgerMetadata newMeta) {
                 if (!metadata.currentEnsemble.get(ensembleInfo.bookieIndex)
                         .equals(ensembleInfo.addr)) {
                     return updateMetadataIfPossible(newMeta);
+                } else {
+                    // it happens when ensemble has multiple concurrent failed bookie change. eg: ensemble [b1,b2]
+                    // First failed bookie [b1] has updated metadata successfully and concurrent second failed bookie
+                    // [b2] call saw conflict zk-metadata version and failed. so, current metadata still has failed
+                    // second bookie [b2] in ensebleInfo. in this case, handle failed bookie by retrying to replace it
+                    // again.
+                    blockAddCompletions.decrementAndGet();
+                    handleBookieFailure(ensembleInfo.addr, ensembleInfo.bookieIndex);
```

Comment:
what thread does this run in? From looking at the blockAddCompletions stuff recently, it only works because all accesses are actually from the same thread (the thread the ordered executor gives you for ledgerId). If this is on another thread, this decrement and call should be pushed onto the ordered thread.

Reply:
> what thread does this run in? If this is on another thread, this decrement and call should be pushed onto the ordered thread.

Yes, it happens on ordered thread. When write fails , that thread tries to update ensembles on another ordered thread and if it sees conflict then that ordered comes here and decrements `blockAddCompletions`.

> Ah, yes, we should push this fix to all non-master branches. For master, the conflict resolution will be completely removed.

Then is it fine, we can still push it tp master for now so, anyone can cherry-pick this change in appropriate branch and eventually this will be removed from master by the new change.?